src/Filters/filt_order.jl

```diff
@@ -0,0 +1,458 @@
+                m2 = m1
+                fm2 = fm1
+                m1 = xn
+                fm1 = fn
+            elseif (fn ≤ fm2 || m2 == m || m2 == m1)
+                m2 = xn
+                fm2 = fn
+            end
+        end
+    end
+    m
+end
+
+#
+# Bandstop cost functions and passband minimization
+#
+for filt in (:butterworth, :elliptic, :chebyshev)
+    @eval begin
+        function $(Symbol(string(filt, "_bsfcost")))(Wx::Real, uselowband::Bool, Wp::Tuple{Real, Real}, Ws::Tuple{Real, Real}, Rp::Real, Rs::Real)
+            # override one of the passband edges with the test frequency.
+            Wpc = uselowband ? tuple(Wx, Wp[2]) : tuple(Wp[1], Wx)
+    
+            # get the new warp frequency.
+            warp = minimum(abs.((Ws .* (Wpc[1]-Wpc[2])) ./ (Ws.^2 .- (Wpc[1]*Wpc[2]))))
+
+            # use the new frequency to determine the filter order.
+            $(Symbol(string(filt, "_order_estimate")))(Rp, Rs, warp)
+        end
+
+        function $(Symbol(string(filt, "_bsfmin")))(Wp::Tuple{Real, Real}, Ws::Tuple{Real, Real}, Rp::Real, Rs::Real)
+            # NOTE: the optimization function will adjust the corner frequencies in Wp, returning a new passband tuple.
+            Δ = eps(typeof(Wp[1]))^(2/3)
+            C₁(w) = $(Symbol(string(filt, "_bsfcost")))(w, true, Wp, Ws, Rp, Rs)
+            p1 = brent(C₁, Wp[1], Ws[1]-Δ)
+
+            C₂(w) = $(Symbol(string(filt, "_bsfcost")))(w, false, tuple(p1, Wp[2]), Ws, Rp, Rs)
+            p2 = brent(C₂, Ws[2]+Δ, Wp[2])
+            Wadj = tuple(p1, p2)
+
+            Wa = (Ws .* (Wadj[1]-Wadj[2])) ./ (Ws.^2 .- (Wadj[1]*Wadj[2]))
+            minimum(abs.(Wa)), Wadj
+        end
+    end
+end
+
+"""
+    (N, ωn) = buttord(Wp::Tuple{Real,Real}, Ws::Tuple{Real,Real}, Rp::Real, Rs::Real; domain=:z)
+
+Butterworth order estimate for bandpass and bandstop filter types.  `Wp` and `Ws` are 2-element pass 
+and stopband frequency edges, with no more than `Rp` dB passband ripple and at least `Rs` dB stopband 
+attenuation. Based on the ordering of passband and bandstop edges,  the Bandstop or Bandpass filter 
+type is inferred. `N` is an integer indicating the lowest estimated filter order, with `ωn` specifying
+the cutoff or "-3 dB" frequencies. If a domain of `:s` is specified, the passband and stopband frequencies 
+are interpretted as radians/second, giving an order and natural frequencies for an analog filter. The default
+domain is `:z`, interpretting the input frequencies as normalized from 0 to 1, where 1 corresponds to π radians/sample.
+"""
+function buttord(Wp::Tuple{Real,Real}, Ws::Tuple{Real,Real}, Rp::Real, Rs::Real; domain::Symbol=:z)
+
+    # make sure the band edges are in increasing order.
+    Wps = (Wp[1] > Wp[2]) ? tuple(Wp[2], Wp[1]) : tuple(Wp[1], Wp[2])
+    Wss = (Ws[1] > Ws[2]) ? tuple(Ws[2], Ws[1]) : tuple(Ws[1], Ws[2])
+
+    # infer filter type based on ordering of edges.
+    ftype = (Wps[1] < Wss[1]) ? Bandstop : Bandpass
+
```

Comment:
Should we verify that `(Wps[1] < Wss[1]) == (Wps[2] > Wss[2])` and throw an informative error otherwise?

Reply:
```suggestion
    (Wps[1] < Wss[1]) != (Wps[2] > Wss[2]) && error("Pass and stopband edges must be ordered for Bandpass/Bandstop cases.")

    # infer filter type based on ordering of edges.
    ftype = (Wps[1] < Wss[1]) ? Bandstop : Bandpass
```
Would this be a suitable enough error message?

Reply:
```julia
(Wps[1] < Wss[1]) != (Wps[2] > Wss[2]) && error("Pass and stopband edges must be ordered for Bandpass/Bandstop filters.")
```
Would this be a contextual enough error message?

Reply:
I don't worry too much about the exact wording. But should probably be an `ArgumentError`.

Reply:
Added! I also included some test cases for `brent` so that we're covered.